mono/metadata/class.c

```diff
@@ -515,6 +521,18 @@ mono_type_get_name_recurse (MonoType *type, GString *str, gboolean is_recursed,
 			MonoTypeNameFormat nested_format;
 			int i;
 
+			// https://bugzilla.xamarin.com/show_bug.cgi?id=57938
+			// Generic instances containing generic parameters from themselves
+			//  should have a null fullname unless they are generic type definitions
+			if (!mono_class_is_gtd (klass)) {
+				for (i = 0; i < inst->type_argc; i++) {
```

Comment:
I wonder if this could be:
```
  if (mono_metadata_type_equal (&klass->byval_arg, mono_class_get_canonical_inst (gclass->container_class)))
    return;
```